lib/CL/devices/remote/network_discovery.c

```diff
@@ -0,0 +1,656 @@
+#include <avahi-common/domain.h>
+#include <avahi-common/error.h>
+#include <avahi-common/malloc.h>
+#include <avahi-common/thread-watch.h>
+
+AvahiThreadedPoll *avahi_threaded_poll = NULL;
+AvahiClient *avahi_client = NULL;
+cl_int domain_count = 0;
+AvahiServiceBrowser **service_browser = NULL;
+#endif
+
+#if defined(ENABLE_REMOTE_DISCOVERY_DHT)
+#include <opendht/opendht_c.h>
+#endif
+
+#define FREE_INFO(__OBJ__)                                                    \
+  do                                                                          \
+    {                                                                         \
+      POCL_MEM_FREE (__OBJ__->id);                                            \
+      POCL_MEM_FREE (__OBJ__->domain);                                        \
+      POCL_MEM_FREE (__OBJ__->ip_port);                                       \
+    }                                                                         \
+  while (0)
+
+/**
+ * This structure contains remote server inforamtion about remote server found
+ * through discovery. A hash table handle (`UT_hash_handle`) is included to
+ * allow instances of this struct to be managed in a hash table, preventing
+ * redundant entries when the same server is discovered multiple times.
+ *
+ * \brief Represents information about a remote server discovered on the
+ * network.
+ */
+typedef struct server_info_t
+{
+  char *id;             // Unique ID with which server advertises itself
+  char *domain;         // Domain name in which the server was found
+  char *ip_port;        // "IP:port"
+  cl_uint device_count; // Number of devices in the remote server
+  UT_hash_handle hh;
+} server_info;
+
+/* Function pointer for the callback defined in pocl runtime to dynamically
+ * add and initialize devices in the discovered remote server.
+ */
+cl_int (*disco_dev_init) (const char *, unsigned);
+
+/* Function pointer for the callback defined in pocl-r driver to reconnect
+ * to a know server with same session.
+ */
+cl_int (*reconnect) (const char *);
+
+unsigned dev_type_idx;
+
+static server_info *server_table = NULL;
+static pocl_lock_t server_info_lock = POCL_LOCK_INITIALIZER;
+
+void
+destroy_server_info_table (server_info *info)
+{
+  server_info *current;
+  server_info *temp;
+
+  HASH_ITER (hh, info, current, temp)
+  {
+    FREE_INFO (current);
+    HASH_DEL (info, current);
+    POCL_MEM_FREE (current);
+  }
+}
+
+/**
+ * Function called by the resolver to add a new server. This function calls the
+ * callback that dynamically adds the devices to the platform.
+ */
+static void
+register_server (const char *id,
+                 const char *domain,
+                 const char *server_key,
+                 cl_uint device_count,
+                 server_info *p_info)
+{
+  if (!device_count)
+    {
+      POCL_MSG_ERR ("Resolver called with zero devices.\n");
+      return;
+    }
+
+  server_info *new_server = p_info;
+  if (!new_server)
```

Comment:
Same here.
```suggestion
  if (new_server == NULL)
```
Strictly, I think NULL is allowed to be non-zero, although this is not likely the case in any existing systems.